lib/irb.rb

```diff
@@ -702,6 +732,14 @@ def inspect
       format("#<%s: %s>", self.class, ary.join(", "))
     end
 
+    def is_assignment_expression?(line)
+      s = Ripper.sexp(line)
+      return false if !s.is_a?(Array)
+      return false if !s[1].is_a?(Array)
+      return false if !s[1][0].is_a?(Array)
+      ASSIGNMENT_NODE_TYPES.include?(s[1][0][0])
```

Comment:
It doesn't work for multi-statement, e.g.,`nil; a = 0`.
And you can use `Array#dig`:
```ruby
ASSIGNMENT_NODE_TYPES.include?(Ripper.sexp(line)&.dig(1, -1, 0))
```

Reply:
I realize it doesn't work for something like `nil; a = 0`. I wasn't necessarily trying to solve it for all cases, just most.

The only problem with [`Array.dig`](https://ruby-doc.org/core-2.3.0/Array.html#method-i-dig) is that if any intermediate element isn't an array (or has its own `dig` method) it raises an error. I wasn't sure if I could rely on the object structure returned from `Ripper.sexp` always being an array whose second element is an array whose first element is an array

Reply:
The logic now looks at the last expression in the array. I still wasn't 100% confident that I could rely on everything being arrays, so I didn't change to `dig`. If it's a guarantee that any non-nil value returned by `Ripper.sexp` can be dug with 3 integer arguments, then I'd make the change.

Reply:
An S-expression is a list in common, and a list is an Array in Ruby.
Even if it would change, it should be an array-like object.
Otherwise you can't access the contents using `[]`.

Reply:
The point is, I don't try to access it with `[]` until after I check that it's an array. I know that at the top level it's an array, and it's all nested arrays, but how do I know that I won't run into a terminal element at some point in the middle of the dig recursion? For example, this would fail: `[1, 2, 3].dig(1, -1, 0)`.

Reply:
The outline of Ripper S-exp hasn't change for a long time.
If the structure changes, the code should have to change too.

Another way is massign.
```ruby
_, (*, (e,)) = Ripper.sexp(line)
ASSIGNMENT_NODE_TYPES.include?(e)
```

Reply:
> The outline of Ripper S-exp hasn't change for a long time.
If the structure changes, the code should have to change too.

Right, but I don't know what that structure is, and if the current structure is guaranteed to always have an array for the last element of the second element of the array returned by `Ripper.sexp` for every possible valid ruby expression. It holds for every valid expression I've tried, but I don't know if what I tried was exhaustive.

If you say it is guaranteed at the current time, then yeah, I think using `&.dig(1,-1,0)` is good.

Reply:
Using an `massign` seems more forgiving than `dig` when the structure doesn't match what one expects. It seems to assign `nil` to variables in cases that `dig` would raise errors. But the code looks more confusing.

If we can be certain that `Ripper.sexp(line)&.dig(1,-1,0)` will never raise an error, then I'm fine going with that. If it's possible that it might raise an error, I'd prefer to leave to code as is so it's more obvious what's going on rather than going with the `massign` expression.

Reply:
The results of ripper will be kept compatible as possible, but it is tough to guarantee.
Although this part would change rarely, I understand your concern.
Feel free to choice the method.

Reply:
I'll use the dig. I used a large Ruby code base as a source of valid Ruby expressions, and every one resulted in a sexp that matched the structure needed by that call to dig.

For the record, I wasn't worried about the sexp structure changing in later versions. I was worried that there might be valid expressions for which the *current* version of ripper would return structures that wouldn't work for the exact arguments to dig from above.